File: src/application/default/change-log/entities/sqlite/File/file.controller.ts
import { Selectable } from "kysely";
import { IChangeLog } from "../../../change-log.model";
import { IFile } from "./file.model";
import { IProjectEntityBaseController } from "../../../../../common/repositories/project-entity.repository";
import { ChangeHistory } from "../../../../../common/types/entity";

export class FileController implements IProjectEntityBaseController {
  validate(changeLog: Selectable<IChangeLog>): void {

    if (!changeLog.entitySchema || typeof changeLog.entitySchema !== "object") {
      const newEntitySchema: Selectable<IFile> = {
        id: 0,
        global_id: 0,
        approval_status: true,
        description: "",
        file_url: "",
        file_local_path: "",
        change_history: {
          user: 0,
          changeType: "create",
          description: "",
          timestamp: new Date(),
          approvalHistory: []
        }
      };

      changeLog.entitySchema = newEntitySchema;
    }

    const entitySchema = changeLog.entitySchema as Selectable<IFile>;

    if (typeof entitySchema.global_id !== "number") {
      entitySchema.global_id = 0;
    }

    if (typeof entitySchema.approval_status !== "boolean") {
      entitySchema.approval_status = true;
    }

    if (typeof entitySchema.description !== "string") {
      entitySchema.description = "";
    }

    if (typeof entitySchema.file_url !== "string") {
      entitySchema.file_url = "";
    }

    if (typeof entitySchema.file_local_path !== "string") {
      entitySchema.file_local_path = "";
    }

    if (typeof entitySchema.change_history !== "object" || entitySchema.change_history === null) {
      entitySchema.change_history = {
        user: 0,
        changeType: "create",
        description: "",
        timestamp: new Date(),
        approvalHistory: []
      } as ChangeHistory;
    }

    changeLog.entitySchema = entitySchema;
  }
}

File: src/controllers/file_controller.cpp
#include "file_controller.h"
#include "common/repository_locator.h"
#include <QDir>

extern std::shared_ptr<User> gUser;
extern int gTenantId;
extern int gProjectId;

FileController::FileController(QObject *parent)
    : QObject{parent},
    m_fileRepository(RepositoryLocator::instance().fileRepository()),
    m_draftEntityRepository(RepositoryLocator::instance().draftEntityRepository())
{}

void FileController::create(const QString &description, const QString &fileUrl, const QString &fileLocalPath) const
{

    /***************** DraftEntity Start *****************/

    QJsonObject jsonObject;
    jsonObject["description"] = description;
    jsonObject["fileUrl"] = fileUrl;
    jsonObject["fileLocalPath"] = fileLocalPath;

    QJsonDocument jsonDoc(jsonObject);
    QString entitySchema = jsonDoc.toJson(QJsonDocument::Indented);

    qDebug() << "File::EntitySchema: " << entitySchema;

    // change history
    QDateTime currentDateTimeUtc = QDateTime::currentDateTimeUtc();
    QString isoTime = currentDateTimeUtc.toString(Qt::ISODateWithMs);

    QJsonObject jsonObjectChangeHistory;
    jsonObjectChangeHistory["user"] = gUser->getId();
    jsonObjectChangeHistory["timestamp"] = isoTime;
    jsonObjectChangeHistory["changeType"] = "create";
    jsonObjectChangeHistory["description"] = "Created File Entity";
    jsonObjectChangeHistory["approvalHistory"] = "null";

    QJsonDocument jsonDocChangeHistory(jsonObjectChangeHistory);
    QString changeHistory = jsonDocChangeHistory.toJson(QJsonDocument::Indented);

    DraftEntity draftEntity;
    draftEntity.setId(0);
    draftEntity.setTenant(gTenantId);
    draftEntity.setCreatedOn(QDate::currentDate());
    draftEntity.setProject(gProjectId);
    draftEntity.setEntity("File");
    draftEntity.setCreatedByUser(gUser->getId());
    draftEntity.setNextApprovingUser(0);
    draftEntity.setEntitySchema(entitySchema);
    draftEntity.setAssociatedApprovedEntity(0);
    draftEntity.setChangeHistory(changeHistory);

    m_draftEntityRepository->saveQML(&draftEntity);
}

void FileController::approvedCreate(const QString &description, const QString &fileUrl, const QString &fileLocalPath) const
{
    File file;

    file.setId(0);
    file.setGlobalId("123");
    file.setApprovalStatus(true);

    file.setDescription(description);
    file.setFileUrl(fileUrl);
    file.setFileLocalPath(fileLocalPath);

    m_fileRepository->saveQML(&file);
}

std::vector<File*> FileController::getFileList(bool isApproved) const
{
    qDebug() << "IsApproved: " << isApproved;

    if (isApproved) {
        return m_fileRepository->findAllQML();
    } else {
        std::vector<DraftEntity*> drafts = m_draftEntityRepository->findAllQML("File");
        std::vector<File*> files;

        for (int i = 0; i < drafts.size(); i++) {

            QString jsonString = drafts[i]->getEntitySchema();
            QJsonDocument jsonDoc = QJsonDocument::fromJson(jsonString.toUtf8());

            if (!jsonDoc.isNull() && jsonDoc.isObject()) {

                auto file = new File();
                QJsonObject jsonObj = jsonDoc.object();

                file->setId(i + 1);
                file->setGlobalId("draft");
                file->setApprovalStatus(false);

                file->setDescription(jsonObj["description"].toString());
                file->setFileUrl(jsonObj["fileUrl"].toString());
                file->setFileLocalPath(jsonObj["fileLocalPath"].toString());

                files.push_back(file);
            }
        }
        return files;
    }
}

QString FileController::copyFileToFusionHub(const QString &sourcePath) const
{
    QString targetDir = "C:/Users/RIPL/Documents/FusionHubData/NMBPL/";

    QDir dir;
    if (!dir.exists(targetDir)) {
        dir.mkpath(targetDir);
    }

    QFileInfo fileInfo(sourcePath);
    QString fileName = fileInfo.fileName();
    QString targetPath = targetDir + "/" + fileName;

    // Overwrite if exists
    if (QFile::exists(targetPath)) {
        QFile::remove(targetPath);
    }

    // Copy file
    if (QFile::copy(sourcePath, targetPath)) {
        qDebug() << "File copied to:" << targetPath;
        return targetPath;
    } else {
        qDebug() << "File copy failed:" << sourcePath;
        return "";
    }
}


